include/zephyr/mgmt/mcumgr/callbacks.h

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef H_MCUMGR_CALLBACKS_
+#define H_MCUMGR_CALLBACKS_
+
+#include <inttypes.h>
+#include <zephyr/sys/slist.h>
+
+#ifdef CONFIG_MCUMGR_CMD_FS_MGMT
+#include <zephyr/mgmt/mcumgr/fs_mgmt_callbacks.h>
+#endif
+
+#ifdef CONFIG_MCUMGR_CMD_IMG_MGMT
+#include <zephyr/mgmt/mcumgr/img_mgmt_callbacks.h>
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief MCUmgr callback API
+ * @defgroup mcumgr_callback_api MCUmgr callback API
+ * @ingroup mcumgr
+ * @{
+ */
+
+/**
+ * @typedef mgmt_cb
+ * @brief Function to be called on MGMT notification/event.
+ *
+ * This callback function is used to notify an application or system about a mcumgr mgmt event.
+ *
+ * @param event		MGMT_EVT_OP_[...].
+ * @param failed	Will be set to true if a previous handler has already returned a failure
+ *			error code (i.e. this handler is being called for a notification only,
+ *			the return code will be ignored)
+ * @param abort_more	Set to true to abort further processing by additional handlers.
+ * @param data		Optional event argument.
+ *
+ * @return		MGMT_ERR_[...] of the status to return to the calling code (only checked
+ *			when failed is false).
+ */
+typedef int (*mgmt_cb)(uint16_t event, bool failed, bool *abort_more, void *data);
+
+/**
+ * MGMT event opcodes.
+ */
+enum {
+	/** Callback when a command is received, data is mgmt_evt_op_cmd_recv_arg. */
+	MGMT_EVT_OP_CMD_RECV			= 1,
+
+	/** Callback when a a status is updated, data is mgmt_evt_op_cmd_status_arg. */
+	MGMT_EVT_OP_CMD_STATUS,
+
+	/** Callback when a command has been processed, data is mgmt_evt_op_cmd_status_arg. */
+	MGMT_EVT_OP_CMD_DONE,
+
+	/** Callback when a file has been accessed in fs_mgmt, data is fs_mgmt_file_access. */
+	MGMT_EVT_OP_FS_MGMT_FILE_ACCESS,
+
+	/**
+	 * Callback when a client sends a file upload chunk in img_mgmt, data is
+	 * img_mgmt_upload_check.
+	 */
+	MGMT_EVT_OP_IMG_MGMT_DFU_CHUNK,
+
+	/** Callback when a DFU operation is stopped in img_mgmt. */
+	MGMT_EVT_OP_IMG_MGMT_DFU_STOPPED,
+
+	/** Callback when a DFU operation is started in img_mgmt. */
+	MGMT_EVT_OP_IMG_MGMT_DFU_STARTED,
+
+	/** Callback when a DFU operation has finished being transferred in img_mgmt. */
+	MGMT_EVT_OP_IMG_MGMT_DFU_PENDING,
+
+	/** Callback when an image has been confirmed in img_mgmt. */
+	MGMT_EVT_OP_IMG_MGMT_DFU_CONFIRMED,
+
+	/** Callback when a reset command has been received in os_mgmt. */
+	MGMT_EVT_OP_OS_MGMT_RESET,
+
+	/** Start ID to be used for custom user events. */
+	MGMT_EVT_OP_USER_CUSTOM_START,
```

Comment:
Right I get what you mean about combining the ID, I will rework it. I don't understand the part about being used outside though, the events would be compiled and used for an application so the order would not matter if it changed (which it should not do anyway, new elements would be appended), or is that for the use case of having an event handler which outputs the callback details out to e.g. a UART for an external application to parse?

Reply:
Can we gut and repurpose these IDs since they are unimplemented in the zephyr version?
```
#define MGMT_GROUP_ID_CONFIG    3
#define MGMT_GROUP_ID_CRASH     5
#define MGMT_GROUP_ID_SPLIT     6
#define MGMT_GROUP_ID_RUN       7
```
Would need 2 IDs, one for base SMP and other for all groups, alternatively could add them as 62 and 63 but seems like these unused IDs could be reused instead.

Reply:
> Can we gut and repurpose these IDs since they are unimplemented in the zephyr version?
> 
> ```
> #define MGMT_GROUP_ID_CONFIG    3
> #define MGMT_GROUP_ID_CRASH     5
> #define MGMT_GROUP_ID_SPLIT     6
> #define MGMT_GROUP_ID_RUN       7
> ```
> 
> Would need 2 IDs, one for base SMP and other for all groups, alternatively could add them as 62 and 63 but seems like these unused IDs could be reused instead.

I am reluctant to remove them. I am quite sure that we will not be using the MGMT_GROUP_ID_SPLIT , but not so sure about others.
Maybe we just need to shift all groups by 4 (3), to reserve 15 (8) first codes for internal stuff like SMP, MGT, transports(?), and leave the rest of space for groups.
I do not think that we will use more than 255 points per event source, so we have more than 16 bits to allocate for these sources (groups, transports, internals).